File: NetDeck/LargeCardCell.m
//
//  LargeCardCell.m
//  Net Deck
//
//  Created by Gereon Steffens on 24.12.13.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "LargeCardCell.h"
#import "CGRectUtils.h"

@interface LargeCardCell()
@property NSArray* pips;
@end
@implementation LargeCardCell

-(void) awakeFromNib
{
    [super awakeFromNib];
    
    self.pips = @[ self.pip1, self.pip2, self.pip3, self.pip4, self.pip5 ];
    
    static int diameter = 8;
    for (UIView* pip in self.pips)
    {
        pip.frame = CGRectSetSize(pip.frame, diameter, diameter);
        pip.layer.cornerRadius = diameter/2;
    }
    
    self.name.font = [UIFont monospacedDigitSystemFontOfSize:15 weight:UIFontWeightMedium];
    
    for (UILabel* lbl in @[ self.label1, self.label2, self.label3]) {
        lbl.font = [UIFont monospacedDigitSystemFontOfSize:17 weight:UIFontWeightRegular];
    }
}

-(void) prepareForReuse {
    for (UIView* pip in self.pips) {
        pip.layer.borderWidth = 0;
    }
}

-(void) setCardCounter:(CardCounter *)cc
{
    [super setCardCounter:cc];
    Card* card = cc.card;
    
    if (card.type == NRCardTypeIdentity)
    {
        self.name.text = card.name;
    }
    else if (card.unique)
    {
        self.name.text = [NSString stringWithFormat:@"%lu× %@ •", (unsigned long)cc.count, card.name];
    }
    else
    {
        self.name.text = [NSString stringWithFormat:@"%lu× %@", (unsigned long)cc.count, card.name];
    }
    
    self.name.textColor = [UIColor blackColor];
    if (!self.deck.isDraft && card.owned < cc.count)
    {
        self.name.textColor = [UIColor redColor];
    }
    
    NSString* factionName = [Faction name:card.faction];
    NSString* typeName = [CardType name:card.type];
    
    NSString* subtype = card.subtype;
    if (subtype.length > 0)
    {
        self.type.text = [NSString stringWithFormat:@"%@ · %@: %@", factionName, typeName, card.subtype];
    }
    else
    {
        self.type.text = [NSString stringWithFormat:@"%@ · %@", factionName, typeName];
    }

    if (self.deck) {
        [self setInfluence:[self.deck influenceFor:cc] andCard:cc];
    } else {
        [self setInfluence:card.influence andCard:cc];
    }
    
    self.copiesLabel.hidden = card.type == NRCardTypeIdentity;
    self.copiesStepper.hidden = card.type == NRCardTypeIdentity;
    self.identityButton.hidden = card.type != NRCardTypeIdentity;
    self.type.hidden = NO;
    
    self.label2.textColor = [UIColor blackColor];
    // labels from top: cost/strength/mu
    switch (card.type)
    {
        case NRCardTypeIdentity:
            self.label1.text = [@(card.minimumDecksize) stringValue];
            self.icon1.image = [ImageCache cardIcon];
            if (card.influenceLimit == -1)
            {
                self.label2.text = @"∞";
            }
            else
            {
                NSInteger inf = self.deck.influenceLimit;
                self.label2.text = [NSString stringWithFormat:@"%ld", (long)inf];
                if (inf != card.influenceLimit) {
                    self.label2.textColor = [UIColor redColor];
                }
            }
            self.icon2.image = [ImageCache influenceIcon];
            if (card.role == NRRoleRunner)
            {
                self.label3.text = [NSString stringWithFormat:@"%ld", (long)card.baseLink];
                self.icon3.image = [ImageCache linkIcon];
            }
            else
            {
                self.label3.text = @"";
                self.icon3.image = nil;
            }
            if (cc == nil)
            {
                self.type.hidden = YES;
                self.label1.text = @"";
                self.label2.text = @"";
                self.label3.text = @"";
                self.icon1.image = nil;
                self.icon2.image = nil;
                self.icon3.image = nil;
                [self.identityButton setTitle:l10n(@"Choose Identity") forState:UIControlStateNormal];
            }
            else
            {
                [self.identityButton setTitle:l10n(@"Switch Identity") forState:UIControlStateNormal];
            }
            break;
            
        case NRCardTypeProgram:
        case NRCardTypeResource:
        case NRCardTypeEvent:
        case NRCardTypeHardware: {
            NSString* cost = card.costString;
            NSString* str = card.strengthString;
            self.label1.text = cost;
            self.icon1.image = cost.length > 0 ? [ImageCache creditIcon] : nil;
            self.label2.text = str;
            self.icon2.image = str.length > 0 ? [ImageCache strengthIcon] : nil;
            self.label3.text = card.mu != -1 ? [NSString stringWithFormat:@"%ld", (long)card.mu] : @"";
            self.icon3.image = card.mu != -1 ? [ImageCache muIcon] : nil;
            break;
        }
            
        case NRCardTypeIce: {
            NSString* cost = card.costString;
            NSString* str = card.strengthString;
            self.label1.text = cost;
            self.icon1.image = cost.length > 0 ? [ImageCache creditIcon] : nil;
            self.label2.text = @"";
            self.icon2.image = nil;
            self.label3.text = str;
            self.icon3.image = str.length > 0 ? [ImageCache strengthIcon] : nil;
            break;
        }
            
        case NRCardTypeAgenda:
            self.label1.text = [NSString stringWithFormat:@"%ld", (long)card.advancementCost];
            self.icon1.image = [ImageCache difficultyIcon];
            self.label2.text = @"";
            self.icon2.image = nil;
            self.label3.text = [NSString stringWithFormat:@"%ld", (long)card.agendaPoints];
            self.icon3.image = [ImageCache apIcon];
            break;
            
        case NRCardTypeAsset:
        case NRCardTypeOperation:
        case NRCardTypeUpgrade: {
            NSString* cost = card.costString;
            self.label1.text = cost;
            self.icon1.image = cost.length > 0 ? [ImageCache creditIcon] : nil;
            self.label2.text = @"";
            self.icon2.image = nil;
            self.label3.text = card.trash != -1 ? [NSString stringWithFormat:@"%ld", (long)card.trash] : @"";
            self.icon3.image = card.trash != -1 ? [ImageCache trashIcon] : nil;
            break;
        }
            
        case NRCardTypeNone:
            NSAssert(NO, @"this can't happen");
            break;
    }
    
    self.copiesStepper.maximumValue = self.deck.isDraft ? 100 : cc.card.maxPerDeck;
    self.copiesStepper.value = cc.count;
    self.copiesLabel.text = [NSString stringWithFormat:@"×%lu", (unsigned long)cc.count];
}

-(void) setInfluence:(NSInteger)influence andCard:(CardCounter*)cc
{
    if (influence > 0)
    {
        self.influenceLabel.textColor = cc.card.factionColor;
        self.influenceLabel.text = [NSString stringWithFormat:@"%ld", (long)influence];
        
        CGColorRef color = cc.card.factionColor.CGColor;
        
        for (int i=0; i<self.pips.count; ++i)
        {
            UIView* pip = self.pips[i];
            pip.layer.backgroundColor = color;
            pip.hidden = i >= cc.card.influence;
        }
    }
    else
    {
        self.influenceLabel.text = @"";
        for (UIView* pip in self.pips)
        {
            pip.hidden = YES;
        }
    }
    
    if (cc.card.isMostWanted) {
        for (UIView* pip in self.pips) {
            // find the first non-hidden pip, and draw it as a black circle
            if (!pip.hidden) {
                continue;
            }
            
            pip.layer.backgroundColor = [UIColor whiteColor].CGColor;
            pip.layer.borderWidth = 1;
            pip.layer.borderColor = [UIColor blackColor].CGColor;
            pip.hidden = NO;
            break;
        }
    }
}

@end

File: NetDeck/LargeCardCell.swift
//
//  LargeCardCell.swift
//  NetDeck
//
//  Created by Gereon Steffens on 31.12.16.
//  Copyright © 2018 Gereon Steffens. All rights reserved.
//

import UIKit

class LargeCardCell: CardCell {
    
    @IBOutlet weak var type: UILabel!
    
    @IBOutlet weak var label1: UILabel!
    @IBOutlet weak var label2: UILabel!
    @IBOutlet weak var label3: UILabel!
    @IBOutlet weak var icon1: UIImageView!
    @IBOutlet weak var icon2: UIImageView!
    @IBOutlet weak var icon3: UIImageView!
    
    @IBOutlet weak var copiesLabel: UILabel!
    
    @IBOutlet weak var pip1: UIView!
    @IBOutlet weak var pip2: UIView!
    @IBOutlet weak var pip3: UIView!
    @IBOutlet weak var pip4: UIView!
    @IBOutlet weak var pip5: UIView!
    @IBOutlet weak var pip6: UIView!
    
    private var pips = [UIView]()
    private var labels = [UILabel]()
    private var icons = [UIImageView]()
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.pips = [ self.pip1, self.pip2, self.pip3, self.pip4, self.pip5, self.pip6 ]
        self.labels = [ self.label1, self.label2, self.label3 ]
        self.icons = [ self.icon1, self.icon2, self.icon3 ]

        self.pips.forEach { $0.layer.cornerRadius = $0.frame.width / 2 }
        
        let font = UIFont.monospacedDigitSystemFont(ofSize: 17, weight: UIFont.Weight.regular)
        self.name.font = font
        self.labels.forEach { $0.font = font }
        
        self.prepareForReuse()
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        
        self.pips.forEach {
            $0.layer.borderWidth = 0
            $0.isHidden = true
        }
        self.labels.forEach { $0.text = nil }
        self.icons.forEach { $0.image = nil }
        self.influenceLabel.text = ""
        self.copiesLabel.text = ""
        self.copiesLabel.textColor = .black
        self.name.textColor = .black
        self.label2.textColor = .black
        self.type.text = ""
        self.type.isHidden = false
    }
    
    override func setCardCounter(_ cardCounter: CardCounter?) {
        super.setCardCounter(cardCounter)
        guard let cc = cardCounter else {
            return
        }
        
        let card = cc.card
        
        self.name.text = cc.displayName(self.deck.mwl)
        
        if !self.deck.isDraft && (cc.count > card.owned || card.isRotated) {
            self.name.textColor = .red
        }
        if self.deck.legality == .cacheRefresh && card.isCore && cc.count > card.quantity {
            self.name.textColor = .red
        }
        if card.banned(self.deck.mwl) {
            self.name.textColor = .red
        }
        
        let factionName = Faction.name(for: card.faction)
        let typeName = CardType.name(for: card.type)
        
        let subtype = card.subtype
        if subtype.count > 0 {
            self.type.text = String(format: "%@ · %@: %@", factionName, typeName, card.subtype)
        } else {
            self.type.text = String(format: "%@ · %@", factionName, typeName)
        }
        
        let influence = self.deck.influenceFor(cc)
        
        if influence > 0 || card.mwlPenalty(self.deck.mwl) > 0 {
            self.influenceLabel.text = influence > 0 ? "\(influence)" : ""
            self.influenceLabel.textColor = cc.card.factionColor
            
            let universalInf = self.deck.universalInfluenceFor(cc)
            LargeCardCell.setInfluencePips(self.pips, influence: influence, universalInfluence: universalInf, count: cc.count, card: card, mwl: self.deck.mwl)
        }
        
        self.copiesLabel.isHidden = card.type == .identity
        self.copiesStepper.isHidden = card.type == .identity
        self.identityButton.isHidden = card.type != .identity
        
        LargeCardCell.setLabels(for: card, labels: self.labels, icons: self.icons)
        if card.type == .identity && !self.deck.isDraft {
            let limit = self.deck.influenceLimit
            self.label2.text = "\(limit)"
            if limit != card.influenceLimit {
                self.label2.textColor = .red
            }
        }

        self.copiesStepper.maximumValue = Double(self.deck.isDraft ? 100 : cc.card.maxPerDeck)
        self.copiesStepper.value = Double(cc.count)
        self.copiesLabel.text = String(format: "×%lu", cc.count)
    }
    
    static func setLabels(for card: Card, labels: [UILabel], icons: [UIImageView]) {
        assert(labels.count == 3)
        assert(icons.count == 3)
        
        switch card.type {
        case .identity:
            labels[0].text = "\(card.minimumDecksize)"
            icons[0].image = ImageCache.cardIcon
            labels[1].text = card.influenceLimit == -1 ? "∞" : "\(card.influenceLimit)"
            icons[1].image = ImageCache.influenceIcon
            if card.role == .runner {
                labels[2].text = "\(card.baseLink)"
                icons[2].image = ImageCache.linkIcon
            }
        case .program, .resource, .event, .hardware:
            let cost = card.costString
            let str = card.strengthString
            labels[0].text = cost
            icons[0].image = cost.count > 0 ? ImageCache.creditIcon : nil
            labels[1].text = str
            icons[1].image = str.count > 0 ? ImageCache.strengthIcon : nil
            labels[2].text = card.mu != -1 ? "\(card.mu)" : nil
            icons[2].image = card.mu != -1 ? ImageCache.muIcon : nil
        case .ice:
            let cost = card.costString
            let str = card.strengthString
            labels[0].text = cost
            icons[0].image = cost.count > 0 ? ImageCache.creditIcon : nil
            labels[1].text = card.trash != -1 ? "\(card.trash)" : nil
            icons[1].image = card.trash != -1 ? ImageCache.trashIcon : nil
            labels[2].text = str
            icons[2].image = str.count > 0 ? ImageCache.strengthIcon : nil
        case .agenda:
            labels[0].text = "\(card.advancementCost)"
            icons[0].image = ImageCache.difficultyIcon
            labels[2].text = "\(card.agendaPoints)"
            icons[2].image = ImageCache.apIcon
        case .asset, .operation, .upgrade:
            let cost = card.costString
            labels[0].text = cost
            icons[0].image = cost.count > 0 ? ImageCache.creditIcon : nil
            labels[2].text = card.trash != -1 ? "\(card.trash)" : nil
            icons[2].image = card.trash != -1 ? ImageCache.trashIcon : nil
        case .none:
            fatalError("this can't happen")
        }
    }
    
    static func setInfluencePips(_ pips: [UIView], influence: Int, universalInfluence: Int, count: Int, card: Card, mwl: MWL) {
        let uInf = universalInfluence / count
        let inf = max(0, (influence / count) - uInf)
        
        for i in stride(from: 0, to: inf, by: 1) {
            let pip = pips[i]
            pip.layer.backgroundColor = card.factionColor.cgColor
            pip.isHidden = false
        }

        if mwl.universalInfluence {
            let maxPip = min(pips.count, inf + uInf)
            for i in stride(from: inf, to: maxPip, by: 1) {
                circlePip(pips[i])
            }
        } else {
            let inf = max(0, inf)
            let penalty = card.mwlPenalty(mwl)
            if penalty > 0 && inf < pips.count {
                circlePip(pips[inf])
            }
        }
    }
    
    private static func circlePip(_ pip: UIView) {
        pip.layer.backgroundColor = UIColor.white.cgColor
        pip.layer.borderWidth = 1
        pip.layer.borderColor = UIColor.black.cgColor
        pip.isHidden = false
    }
}
